zrml/court/src/lib.rs

```diff
@@ -16,611 +16,2242 @@
+        pub fn necessary_draws_weight(appeals_len: usize) -> usize {
+            // 2^(appeals_len) * 31 + 2^(appeals_len) - 1
+            // MaxAppeals - 1 (= 3) example: 2^3 * 31 + 2^3 - 1 = 255
+            APPEAL_BASIS
+                .saturating_pow(appeals_len as u32)
+                .saturating_mul(INITIAL_DRAWS_NUM)
+                .saturating_add(APPEAL_BASIS.saturating_pow(appeals_len as u32).saturating_sub(1))
         }
 
-        // Every juror that not voted on the first or second most voted outcome are slashed.
+        // Slash the losers and use the slashed amount plus the reward pot to reward the winners.
         fn slash_losers_to_award_winners(
-            valid_winners_and_losers: &[(&T::AccountId, &OutcomeReport)],
-            winner_outcome: &OutcomeReport,
-        ) -> DispatchResult {
-            let mut total_incentives = BalanceOf::<T>::from(0u8);
-            let mut total_winners = BalanceOf::<T>::from(0u8);
-
-            for (jai, outcome) in valid_winners_and_losers {
-                if outcome == &winner_outcome {
-                    total_winners = total_winners.saturating_add(BalanceOf::<T>::from(1u8));
-                } else {
-                    let all_reserved =
-                        CurrencyOf::<T>::reserved_balance_named(&Self::reserve_id(), jai);
-                    // Unsigned division will never overflow
-                    let slash = all_reserved
-                        .checked_div(&BalanceOf::<T>::from(2u8))
-                        .ok_or(DispatchError::Other("Zero division"))?;
-                    CurrencyOf::<T>::slash_reserved_named(&Self::reserve_id(), jai, slash);
-                    total_incentives = total_incentives.saturating_add(slash);
-                }
-            }
+            court_id: CourtId,
+            jurors_to_stakes: BTreeMap<T::AccountId, JurorVoteWithStakesOf<T>>,
+            winner_vote_item: &VoteItem,
+        ) {
+            let mut total_incentives = <NegativeImbalanceOf<T>>::zero();
 
-            let individual_winner_incentive =
-                if let Some(e) = total_incentives.checked_div(&total_winners) {
-                    e
-                } else {
-                    // No winners
-                    return Ok(());
+            let slash_all_delegators =
+                |delegations: &[(T::AccountId, BalanceOf<T>)]| -> NegativeImbalanceOf<T> {
+                    let mut total_imb = <NegativeImbalanceOf<T>>::zero();
+                    for (delegator, d_slashable) in delegations.iter() {
+                        let (imb, missing) = T::Currency::slash(delegator, *d_slashable);
+                        total_imb.subsume(imb);
+                        debug_assert!(
+                            missing.is_zero(),
+                            "Could not slash all of the amount for delegator {:?}.",
+                            delegator
+                        );
```

Comment:
> Why is it allowed to stake an amount that is not an integer multiple of min_stake?

I assume `min_stake` changes in the future. So if we adjust the code so that every juror needs to stake a multiple of `min_stake`, and we change the configuration (because ZTG appreciates in value), the previous staked value is not necessarily a multiple of the new `min_stake` anymore.

> This can happen if a loser has less than min_stake in their account.

The selection algorithm ensures that the `draw.slashable` is actually a valid balance which exists. 

The `unconsumed` stake is the valid balance in this case, which was not already `consumed` by other court cases. 

https://github.com/zeitgeistpm/zeitgeist/blob/3996a190a624e24e20afd5d63da7cb81f3973bce/zrml/court/src/lib.rs#L1407-L1408